craft_providers/lxd/lxd_instance.py

```diff
@@ -64,6 +78,51 @@ def __init__(
         else:
             self.lxc = lxc
 
+    def _set_instance_name(self) -> None:
+        """Convert a name to a LXD-compatible name.
+
+        LXD naming convention:
+        - be between 1 and 63 characters long
+        - be made up exclusively of letters, numbers and dashes from the ASCII table
+        - not start with a digit or a dash
+        - not end with a dash
+
+        To create a LXD-compatible name, forbidden characters are removed, the name is
+        truncated to 40 characters, then a hash is appended:
+        <truncated-name>-<hash-of-name>
+        └     1 - 40   ┘1└     20     ┘
+
+        :param name: name of instance
+        :raises LXDError: if name contains no alphanumeric characters
+        """
+        # remove anything that is not an alphanumeric characters or dash
+        name_with_valid_chars = re.sub(r"[^\w-]", "", self.name)
+        if name_with_valid_chars == "":
+            raise LXDError(
+                brief=(f"failed to create LXD instance with name {self.name!r}."),
+                details="name must contain at least one alphanumeric character",
+            )
+
+        # trim digits and dashes from the beginning and dashes from the end
+        trimmed_name = re.compile(r"^[0-9-]*(?P<valid_part_of_name>.*?)[-]*$").search(
+            name_with_valid_chars
+        )
+        if not trimmed_name or trimmed_name.group("valid_part_of_name") == "":
+            raise LXDError(
+                brief=(f"failed to create LXD instance with name {self.name!r}."),
+                details="name must contain at least one alphanumeric character",
+            )
+
+        # truncate to 40 characters
+        truncated_name = trimmed_name.group("valid_part_of_name")[:40]
+
+        # hash the entire name, not the truncated name
+        hasher = hashlib.sha1()
+        hasher.update(self.name.encode())
+        hashed_name = hasher.hexdigest()[:20]
+
+        self.instance_name = f"{truncated_name}-{hashed_name}"
```

Comment:
Shouldn't we use the new instance name only if the old-style name is invalid or too large, to keep using existing instances?

Reply:
Good catch, I added that in.